File: crates/rivet-gui/ui/src/app/tags.rs
fn load_tag_schema() -> TagSchema {
  match toml::from_str::<TagSchema>(
    TAG_SCHEMA_TOML
  ) {
    | Ok(schema)
      if !schema.keys.is_empty() =>
    {
      tracing::info!(
        version = schema.version,
        key_count = schema.keys.len(),
        "loaded tag schema"
      );
      schema
    }
    | Ok(_) => {
      tracing::warn!(
        "tag schema was empty; using \
         fallback schema"
      );
      TagSchema::default()
    }
    | Err(error) => {
      tracing::error!(%error, "failed to parse tag schema; using fallback schema");
      TagSchema::default()
    }
  }
}

fn kanban_columns_from_schema(
  schema: &TagSchema
) -> Vec<String> {
  let values = schema
    .key(KANBAN_TAG_KEY)
    .map(|entry| {
      entry
        .values
        .iter()
        .map(|value| {
          value.trim().to_string()
        })
        .filter(|value| {
          !value.is_empty()
        })
        .collect::<Vec<_>>()
    })
    .unwrap_or_default();

  if values.is_empty() {
    vec![
      "todo".to_string(),
      "working".to_string(),
      "finished".to_string(),
    ]
  } else {
    values
  }
}

fn optional_text(
  text: &str
) -> Option<String> {
  let trimmed = text.trim();
  if trimmed.is_empty() {
    None
  } else {
    Some(trimmed.to_string())
  }
}

fn split_tags(
  text: &str
) -> Vec<String> {
  text
    .split_whitespace()
    .map(str::trim)
    .filter(|value| !value.is_empty())
    .map(ToString::to_string)
    .collect()
}

fn recurrence_from_tags(
  tags: &[String]
) -> (
  String,
  String,
  Vec<String>,
  Vec<String>,
  String
) {
  let pattern = first_tag_value(
    tags,
    RECUR_TAG_KEY
  )
  .map(normalize_recurrence_pattern)
  .unwrap_or_else(|| {
    "none".to_string()
  });

  let time = first_tag_value(
    tags,
    RECUR_TIME_TAG_KEY
  )
  .unwrap_or_default()
  .to_string();

  let days = split_csv_values(
    first_tag_value(
      tags,
      RECUR_DAYS_TAG_KEY
    )
    .unwrap_or_default()
  )
  .into_iter()
  .map(|entry| {
    entry
      .to_ascii_lowercase()
      .trim()
      .to_string()
  })
  .filter(|entry| {
    WEEKDAY_KEYS
      .iter()
      .any(|key| key == entry)
  })
  .collect::<Vec<_>>();

  let months = split_csv_values(
    first_tag_value(
      tags,
      RECUR_MONTHS_TAG_KEY
    )
    .unwrap_or_default()
  )
  .into_iter()
  .map(|entry| {
    entry
      .to_ascii_lowercase()
      .trim()
      .to_string()
  })
  .filter(|entry| {
    MONTH_KEYS
      .iter()
      .any(|key| key == entry)
  })
  .collect::<Vec<_>>();

  let month_day = first_tag_value(
    tags,
    RECUR_MONTH_DAY_TAG_KEY
  )
  .unwrap_or_default()
  .to_string();

  (
    pattern, time, days, months,
    month_day
  )
}

fn normalize_recurrence_pattern(
  value: &str
) -> String {
  match value
    .trim()
    .to_ascii_lowercase()
    .as_str()
  {
    | "daily" => "daily".to_string(),
    | "weekly" => "weekly".to_string(),
    | "months" => "months".to_string(),
    | "monthly" => {
      "monthly".to_string()
    }
    | "yearly" => "yearly".to_string(),
    | _ => "none".to_string()
  }
}

fn split_csv_values(
  value: &str
) -> Vec<String> {
  value
    .split(',')
    .map(str::trim)
    .filter(|entry| !entry.is_empty())
    .map(ToString::to_string)
    .collect()
}

fn is_recurrence_tag(
  tag: &str
) -> bool {
  matches!(
    tag.split_once(':'),
    Some((key, _))
      if key == RECUR_TAG_KEY
      || key == RECUR_TIME_TAG_KEY
      || key == RECUR_DAYS_TAG_KEY
      || key == RECUR_MONTHS_TAG_KEY
      || key == RECUR_MONTH_DAY_TAG_KEY
  )
}

fn append_recurrence_tags(
  tags: &mut Vec<String>,
  state: &ModalState
) {
  remove_tags_for_key(
    tags,
    RECUR_TAG_KEY
  );
  remove_tags_for_key(
    tags,
    RECUR_TIME_TAG_KEY
  );
  remove_tags_for_key(
    tags,
    RECUR_DAYS_TAG_KEY
  );
  remove_tags_for_key(
    tags,
    RECUR_MONTHS_TAG_KEY
  );
  remove_tags_for_key(
    tags,
    RECUR_MONTH_DAY_TAG_KEY
  );

  let pattern =
    normalize_recurrence_pattern(
      &state.recurrence_pattern
    );
  if pattern == "none" {
    return;
  }

  push_tag_unique(
    tags,
    format!(
      "{RECUR_TAG_KEY}:{pattern}"
    )
  );

  let recurrence_time =
    state.recurrence_time.trim();
  if !recurrence_time.is_empty() {
    push_tag_unique(
      tags,
      format!(
        "{RECUR_TIME_TAG_KEY}:\
         {recurrence_time}"
      )
    );
  }

  if pattern == "weekly"
    && !state.recurrence_days.is_empty()
  {
    let values = state
      .recurrence_days
      .iter()
      .map(|value| {
        value
          .trim()
          .to_ascii_lowercase()
      })
      .filter(|value| {
        WEEKDAY_KEYS
          .iter()
          .any(|key| key == value)
      })
      .collect::<Vec<_>>();
    if !values.is_empty() {
      push_tag_unique(
        tags,
        format!(
          "{RECUR_DAYS_TAG_KEY}:{}",
          values.join(",")
        )
      );
    }
  }

  if pattern == "monthly"
    || pattern == "months"
    || pattern == "yearly"
  {
    let months = state
      .recurrence_months
      .iter()
      .map(|value| {
        value
          .trim()
          .to_ascii_lowercase()
      })
      .filter(|value| {
        MONTH_KEYS
          .iter()
          .any(|key| key == value)
      })
      .collect::<Vec<_>>();
    if !months.is_empty() {
      push_tag_unique(
        tags,
        format!(
          "{RECUR_MONTHS_TAG_KEY}:{}",
          months.join(",")
        )
      );
    }

    let month_day =
      state.recurrence_month_day.trim();
    if !month_day.is_empty() {
      push_tag_unique(
        tags,
        format!(
          "{RECUR_MONTH_DAY_TAG_KEY}:\
           {month_day}"
        )
      );
    }
  }
}

fn collect_tags_for_submit(
  state: &ModalState,
  board_tag: Option<String>,
  allow_recurrence: bool,
  ensure_kanban_lane: bool,
  default_kanban_lane: &str
) -> Vec<String> {
  let mut tags =
    state.draft_tags.clone();
  for tag in
    split_tags(&state.draft_custom_tag)
  {
    push_tag_unique(&mut tags, tag);
  }

  remove_tags_for_key(
    &mut tags,
    BOARD_TAG_KEY
  );
  if let Some(tag) = board_tag {
    push_tag_unique(&mut tags, tag);
  }

  if allow_recurrence {
    append_recurrence_tags(
      &mut tags, state
    );
  }

  if ensure_kanban_lane
    && !tags.iter().any(|tag| {
      tag.starts_with(&format!(
        "{KANBAN_TAG_KEY}:"
      ))
    })
  {
    push_tag_unique(
      &mut tags,
      format!(
        "{KANBAN_TAG_KEY}:{}",
        default_kanban_lane
      )
    );
  }

  tags
}

fn push_tag_unique(
  tags: &mut Vec<String>,
  tag: String
) -> bool {
  let trimmed = tag.trim();
  if trimmed.is_empty() {
    return false;
  }

  if tags
    .iter()
    .any(|existing| existing == trimmed)
  {
    return false;
  }

  tags.push(trimmed.to_string());
  true
}

fn first_value_for_key(
  schema: &TagSchema,
  key: &str
) -> Option<String> {
  schema.key(key).and_then(|entry| {
    entry.values.first().cloned()
  })
}

fn is_single_select_key(
  schema: &TagSchema,
  key: &str
) -> bool {
  schema
    .key(key)
    .and_then(|entry| {
      entry.selection.as_deref()
    })
    .is_some_and(|selection| {
      selection
        .eq_ignore_ascii_case("single")
    })
}

fn remove_tags_for_key(
  tags: &mut Vec<String>,
  key: &str
) {
  tags.retain(|existing| {
    match existing.split_once(':') {
      | Some((existing_key, _)) => {
        existing_key != key
      }
      | None => true
    }
  });
}

fn first_tag_value<'a>(
  tags: &'a [String],
  key: &str
) -> Option<&'a str> {
  tags.iter().find_map(|tag| {
    match tag.split_once(':') {
      | Some((existing_key, value))
        if existing_key == key =>
      {
        Some(value)
      }
      | _ => None
    }
  })
}

fn task_has_tag_value(
  tags: &[String],
  key: &str,
  value: &str
) -> bool {
  tags.iter().any(|tag| {
    matches!(
      tag.split_once(':'),
      Some((existing_key, existing_value))
        if existing_key == key
          && existing_value == value
    )
  })
}

fn remove_board_tag_for_id(
  tags: &mut Vec<String>,
  board_id: &str
) {
  tags.retain(|tag| {
    match tag.split_once(':') {
      | Some((key, value)) => {
        !(key == BOARD_TAG_KEY
          && value == board_id)
      }
      | None => true
    }
  });
}

fn tag_chip_style(
  schema: &TagSchema,
  tag: &str,
  tag_colors: &BTreeMap<String, String>
) -> String {
  let Some((key, _value)) =
    tag.split_once(':')
  else {
    return String::new();
  };

  let color = schema
    .key(key)
    .and_then(|entry| {
      entry.color.as_deref()
    })
    .map(ToString::to_string)
    .or_else(|| {
      tag_colors.get(key).cloned()
    })
    .unwrap_or_else(|| {
      deterministic_tag_key_color(key)
    });

  format!("--tag-key-color:{color};")
}

fn deterministic_tag_key_color(
  key: &str
) -> String {
  let mut hash: u32 = 0x811c9dc5;
  for byte in key.as_bytes() {
    hash ^= u32::from(*byte);
    hash = hash.wrapping_mul(16777619);
  }
  let hue = hash % 360;
  format!("hsl({hue} 72% 54%)")
}

fn tag_badge_style(
  tag: &str,
  tag_colors: &BTreeMap<String, String>
) -> String {
  let Some((key, _)) =
    tag.split_once(':')
  else {
    return String::new();
  };

  let color = tag_colors
    .get(key)
    .cloned()
    .unwrap_or_else(|| {
      deterministic_tag_key_color(key)
    });

  format!("--tag-key-color:{color};")
}

fn tag_color_for_schema_key(
  key: &TagKey
) -> Option<String> {
  if let Some(color) =
    key.color.as_ref()
  {
    return Some(color.clone());
  }
  if key.id.trim().is_empty() {
    return None;
  }
  Some(deterministic_tag_key_color(
    &key.id
  ))
}

fn build_tag_color_map(
  schema: &TagSchema
) -> BTreeMap<String, String> {
  schema
    .keys
    .iter()
    .filter_map(|key| {
      tag_color_for_schema_key(key).map(
        |color| (key.id.clone(), color)
      )
    })
    .collect()
}

#[cfg(test)]
mod tags_tests {
  use super::*;

  fn base_modal_state() -> ModalState {
    ModalState {
      mode:                 ModalMode::Add,
      draft_title:          "title".to_string(),
      draft_desc:           String::new(),
      draft_project:        String::new(),
      draft_board_id:       None,
      lock_board_selection: false,
      draft_custom_tag:     String::new(),
      draft_tags:           vec![],
      picker_key:           None,
      picker_value:         None,
      draft_due:            String::new(),
      recurrence_pattern:   "none"
        .to_string(),
      recurrence_time:      String::new(),
      recurrence_days:      vec![],
      recurrence_months:    vec![],
      recurrence_month_day: String::new(),
      error:                None
    }
  }

  #[test]
  fn recurrence_tags_round_trip_weekly()
  {
    let mut state = base_modal_state();
    state.recurrence_pattern =
      "weekly".to_string();
    state.recurrence_time =
      "15:23".to_string();
    state.recurrence_days = vec![
      "mon".to_string(),
      "wed".to_string(),
      "fri".to_string(),
    ];

    let mut tags = vec![
      "area:software".to_string()
    ];
    append_recurrence_tags(
      &mut tags, &state
    );

    let (
      pattern,
      time,
      days,
      months,
      month_day,
    ) = recurrence_from_tags(&tags);
    assert_eq!(pattern, "weekly");
    assert_eq!(time, "15:23");
    assert_eq!(
      days,
      vec![
        "mon".to_string(),
        "wed".to_string(),
        "fri".to_string()
      ]
    );
    assert!(months.is_empty());
    assert!(month_day.is_empty());
  }

  #[test]
  fn collect_tags_for_submit_overwrites_board_and_preserves_existing_lane(
  ) {
    let mut state = base_modal_state();
    state.draft_tags = vec![
      "area:software".to_string(),
      "board:old".to_string(),
      "kanban:working".to_string(),
    ];
    state.draft_custom_tag =
      "topic:rust urgent".to_string();

    let tags = collect_tags_for_submit(
      &state,
      Some("board:ops".to_string()),
      true,
      true,
      "todo",
    );

    assert!(task_has_tag_value(
      &tags, BOARD_TAG_KEY, "ops"
    ));
    assert!(!task_has_tag_value(
      &tags, BOARD_TAG_KEY, "old"
    ));
    assert!(task_has_tag_value(
      &tags, KANBAN_TAG_KEY, "working"
    ));
    assert!(tags.iter().any(|tag| {
      tag == "topic:rust"
    }));
    assert!(tags.iter().any(|tag| {
      tag == "urgent"
    }));
  }

  #[test]
  fn collect_tags_for_submit_adds_default_lane_when_missing(
  ) {
    let state = base_modal_state();
    let tags = collect_tags_for_submit(
      &state, None, true, true,
      "todo"
    );
    assert!(task_has_tag_value(
      &tags, KANBAN_TAG_KEY, "todo"
    ));
  }
}

File: crates/rivet-gui/ui/web/lib/tags.ts
import type { TagKey, TagSchema } from "../types/config";
import type { RecurrenceDraft } from "../types/ui";

export const KANBAN_TAG_KEY = "kanban";
export const BOARD_TAG_KEY = "board";
export const RECUR_TAG_KEY = "recur";
export const RECUR_TIME_TAG_KEY = "recur_time";
export const RECUR_DAYS_TAG_KEY = "recur_days";
export const RECUR_MONTHS_TAG_KEY = "recur_months";
export const RECUR_MONTH_DAY_TAG_KEY = "recur_day";
export const CAL_SOURCE_TAG_KEY = "cal_source";
export const CAL_COLOR_TAG_KEY = "cal_color";
export const CALENDAR_UNAFFILIATED_COLOR = "#7f8691";

export const WEEKDAY_KEYS = ["mon", "tue", "wed", "thu", "fri", "sat", "sun"] as const;
export const MONTH_KEYS = ["jan", "feb", "mar", "apr", "may", "jun", "jul", "aug", "sep", "oct", "nov", "dec"] as const;

export type WeekdayKey = (typeof WEEKDAY_KEYS)[number];
export type MonthKey = (typeof MONTH_KEYS)[number];

export function splitTags(text: string): string[] {
  return text
    .split(/\s+/)
    .map((entry) => entry.trim())
    .filter((entry) => entry.length > 0);
}

export function pushTagUnique(tags: string[], tag: string): boolean {
  const normalized = tag.trim();
  if (normalized.length === 0) {
    return false;
  }
  if (tags.some((entry) => entry === normalized)) {
    return false;
  }
  tags.push(normalized);
  return true;
}

export function removeTagsForKey(tags: string[], key: string): void {
  for (let index = tags.length - 1; index >= 0; index -= 1) {
    const tag = tags[index];
    const [entryKey] = splitTag(tag);
    if (entryKey === key) {
      tags.splice(index, 1);
    }
  }
}

export function splitTag(tag: string): [string | null, string | null] {
  const pair = tag.split(":");
  if (pair.length < 2) {
    return [null, null];
  }
  const key = pair[0]?.trim() ?? "";
  const value = pair.slice(1).join(":").trim();
  if (!key || !value) {
    return [null, null];
  }
  return [key, value];
}

export function firstTagValue(tags: string[], key: string): string | null {
  for (const tag of tags) {
    const [entryKey, entryValue] = splitTag(tag);
    if (entryKey === key && entryValue) {
      return entryValue;
    }
  }
  return null;
}

export function taskHasTagValue(tags: string[], key: string, value: string): boolean {
  return tags.some((tag) => {
    const [entryKey, entryValue] = splitTag(tag);
    return entryKey === key && entryValue === value;
  });
}

export function normalizeHexColor(value: string): string | null {
  const raw = value.trim().replace(/^#/, "");
  if (/^[0-9a-fA-F]{3}$/.test(raw)) {
    const expanded = raw
      .split("")
      .map((ch) => `${ch}${ch}`)
      .join("")
      .toLowerCase();
    return `#${expanded}`;
  }
  if (/^[0-9a-fA-F]{6}$/.test(raw)) {
    return `#${raw.toLowerCase()}`;
  }
  return null;
}

export function normalizeMarkerColor(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    return CALENDAR_UNAFFILIATED_COLOR;
  }
  return normalizeHexColor(trimmed) ?? trimmed;
}

export function deterministicTagKeyColor(key: string): string {
  let hash = 0x811c9dc5;
  for (let index = 0; index < key.length; index += 1) {
    hash ^= key.charCodeAt(index);
    hash = (hash * 16777619) >>> 0;
  }
  const hue = hash % 360;
  return `hsl(${hue} 72% 54%)`;
}

export function colorForTagKey(key: TagKey): string {
  const explicit = key.color?.trim();
  if (explicit) {
    return explicit;
  }
  return deterministicTagKeyColor(key.id);
}

export function buildTagColorMap(schema: TagSchema | null): Record<string, string> {
  const map: Record<string, string> = {};
  for (const key of schema?.keys ?? []) {
    if (!key.id?.trim()) {
      continue;
    }
    map[key.id] = colorForTagKey(key);
  }
  return map;
}

export function tagColorStyle(tag: string, schema: TagSchema | null, colorMap: Record<string, string>): string {
  const [key] = splitTag(tag);
  if (!key) {
    return "";
  }
  const fromSchema = schema?.keys?.find((entry) => entry.id === key)?.color?.trim();
  const color = fromSchema || colorMap[key] || deterministicTagKeyColor(key);
  return color;
}

export function defaultKanbanLane(schema: TagSchema | null): string {
  const laneKey = schema?.keys?.find((entry) => entry.id === KANBAN_TAG_KEY);
  const first = laneKey?.values?.find((entry) => entry.trim().length > 0)?.trim();
  return first || "todo";
}

export function kanbanColumnsFromSchema(schema: TagSchema | null): string[] {
  const laneKey = schema?.keys?.find((entry) => entry.id === KANBAN_TAG_KEY);
  const values = (laneKey?.values ?? []).map((entry) => entry.trim()).filter((entry) => entry.length > 0);
  if (values.length === 0) {
    return ["todo", "working", "finished"];
  }
  return values;
}

export function normalizeRecurrencePattern(value: string): RecurrenceDraft["pattern"] {
  const normalized = value.trim().toLowerCase();
  if (
    normalized === "daily" ||
    normalized === "weekly" ||
    normalized === "months" ||
    normalized === "monthly" ||
    normalized === "yearly"
  ) {
    return normalized;
  }
  return "none";
}

export function appendRecurrenceTags(tags: string[], recurrence: RecurrenceDraft): void {
  removeTagsForKey(tags, RECUR_TAG_KEY);
  removeTagsForKey(tags, RECUR_TIME_TAG_KEY);
  removeTagsForKey(tags, RECUR_DAYS_TAG_KEY);
  removeTagsForKey(tags, RECUR_MONTHS_TAG_KEY);
  removeTagsForKey(tags, RECUR_MONTH_DAY_TAG_KEY);

  const pattern = normalizeRecurrencePattern(recurrence.pattern);
  if (pattern === "none") {
    return;
  }

  pushTagUnique(tags, `${RECUR_TAG_KEY}:${pattern}`);
  const recurrenceTime = recurrence.time.trim();
  if (recurrenceTime.length > 0) {
    pushTagUnique(tags, `${RECUR_TIME_TAG_KEY}:${recurrenceTime}`);
  }

  if (pattern === "weekly") {
    const days = recurrence.days
      .map((day) => day.trim().toLowerCase())
      .filter((day): day is WeekdayKey => WEEKDAY_KEYS.includes(day as WeekdayKey));
    if (days.length > 0) {
      pushTagUnique(tags, `${RECUR_DAYS_TAG_KEY}:${days.join(",")}`);
    }
  }

  if (pattern === "months" || pattern === "monthly" || pattern === "yearly") {
    const months = recurrence.months
      .map((month) => month.trim().toLowerCase())
      .filter((month): month is MonthKey => MONTH_KEYS.includes(month as MonthKey));
    if (months.length > 0) {
      pushTagUnique(tags, `${RECUR_MONTHS_TAG_KEY}:${months.join(",")}`);
    }
    const monthDay = recurrence.monthDay.trim();
    if (monthDay.length > 0) {
      pushTagUnique(tags, `${RECUR_MONTH_DAY_TAG_KEY}:${monthDay}`);
    }
  }
}

export function recurrenceFromTags(tags: string[]): RecurrenceDraft {
  const pattern = normalizeRecurrencePattern(firstTagValue(tags, RECUR_TAG_KEY) ?? "none");
  const time = firstTagValue(tags, RECUR_TIME_TAG_KEY) ?? "";
  const days = (firstTagValue(tags, RECUR_DAYS_TAG_KEY) ?? "")
    .split(",")
    .map((entry) => entry.trim().toLowerCase())
    .filter((entry): entry is WeekdayKey => WEEKDAY_KEYS.includes(entry as WeekdayKey));
  const months = (firstTagValue(tags, RECUR_MONTHS_TAG_KEY) ?? "")
    .split(",")
    .map((entry) => entry.trim().toLowerCase())
    .filter((entry): entry is MonthKey => MONTH_KEYS.includes(entry as MonthKey));
  const monthDay = firstTagValue(tags, RECUR_MONTH_DAY_TAG_KEY) ?? "";
  return {
    pattern,
    time,
    days,
    months,
    monthDay
  };
}

export function isSingleSelectKey(schema: TagSchema | null, keyId: string): boolean {
  const key = schema?.keys?.find((entry) => entry.id === keyId);
  return (key?.selection ?? "").toLowerCase() === "single";
}

export interface CollectSubmitTagsInput {
  selectedTags: string[];
  customTagInput: string;
  boardTag: string | null;
  allowRecurrence: boolean;
  recurrence: RecurrenceDraft;
  ensureKanbanLane: boolean;
  defaultKanbanLaneValue: string;
}

export function collectTagsForSubmit(input: CollectSubmitTagsInput): string[] {
  const tags = [...input.selectedTags];
  for (const tag of splitTags(input.customTagInput)) {
    pushTagUnique(tags, tag);
  }

  removeTagsForKey(tags, BOARD_TAG_KEY);
  if (input.boardTag) {
    pushTagUnique(tags, input.boardTag);
  }

  if (input.allowRecurrence) {
    appendRecurrenceTags(tags, input.recurrence);
  }

  if (input.ensureKanbanLane && !tags.some((tag) => tag.startsWith(`${KANBAN_TAG_KEY}:`))) {
    pushTagUnique(tags, `${KANBAN_TAG_KEY}:${input.defaultKanbanLaneValue}`);
  }
  return tags;
}

export function tagsForKanbanMove(
  tags: string[],
  lane: string,
  boardId?: string | null
): string[] {
  const next = [...tags];
  removeTagsForKey(next, KANBAN_TAG_KEY);
  pushTagUnique(next, `${KANBAN_TAG_KEY}:${lane}`);

  // `undefined` means keep board unchanged; null/"" means clear board.
  if (typeof boardId !== "undefined") {
    removeTagsForKey(next, BOARD_TAG_KEY);
    const normalizedBoardId = boardId?.trim() ?? "";
    if (normalizedBoardId.length > 0) {
      pushTagUnique(next, `${BOARD_TAG_KEY}:${normalizedBoardId}`);
    }
  }

  return next;
}

export function boardIdFromTaskTags(tags: string[]): string | null {
  return firstTagValue(tags, BOARD_TAG_KEY);
}

export function kanbanLaneFromTask(tags: string[], availableLanes: string[], fallbackLane: string): string {
  const lane = firstTagValue(tags, KANBAN_TAG_KEY);
  if (!lane) {
    return fallbackLane;
  }
  if (!availableLanes.includes(lane)) {
    return fallbackLane;
  }
  return lane;
}

export function humanizeLane(value: string): string {
  return value
    .split(/[-_]/g)
    .filter((part) => part.length > 0)
    .map((part) => `${part[0]?.toUpperCase() ?? ""}${part.slice(1).toLowerCase()}`)
    .join(" ");
}
